### PR TITLE
Pass seed to C++ code and use std::default_random_engine

### LINKED ISSUE #71: Isolate RNG per instance and thread
Right now we use cstdlib's `rand` which is not thread-safe.

We should switch to something like `std::default_random_engine generator`.

Related to #30.

### PR DESCRIPTION
**Pull request by hhtong**
Closes #71 and #30

### CODE REVIEW

File: tabu/sampler.py [hunk 1]
```diff
@@ -151,8 +151,12 @@ def sample(self, bqm, initial_states=None, initial_states_generator='random',
 
         # run Tabu search
         samples = np.empty((parsed.num_reads, len(bqm)), dtype=np.int8)
+
+        rng = np.random.RandomState(seed)
```
Comment (randomir) [inline tabu/sampler.py:155]: I'd [switch to `default_rng`](https://numpy.org/doc/stable/reference/random/index.html#quick-start).

File: tabu/src/tabu_search.cc [hunk 2]
```diff
@@ -22,20 +22,14 @@
...
-        throw Exception("length of init_solution doesn't match the size of Q");
-
-    if (tenure < 0 || tenure > (nvars - 1))
-        throw Exception("tenure must be in the range [0, num_vars - 1]");
-    else if (tenure > 0) {
-        tabooTenure = tenure;
-    }
-    else {
-        tabooTenure = (20 < (int)(bqp.nVars / 4.0))? 20 : (int)(bqp.nVars / 4.0);
-    }
```
Comment (randomir) [inline tabu/src/tabu_search.cc:38]: Why remove these checks?

File: tabu/src/tabu_search.cc [hunk 3]
```diff
@@ -117,6 +117,7 @@ double TabuSearch::simpleTabuSearch(const vector<int> &starting,
                                     double startingObjective,
                                     long long ZCoeff,
                                     long long timeLimitInMilliSecs,
+                                    bool useTimeLimit,
```
Comment (randomir) [inline tabu/src/tabu_search.cc:120]: Why introduce this additional argument? (redundant with `timeLimitInMilliSecs`)

File: tests/test_sampler.py [hunk 4]
```diff
@@ -200,3 +205,25 @@ def test_soft_num_reads(self):
 
         # if explicitly given, without initial_states, they are generated
         self.assertEqual(len(tabu.TabuSampler().sample(bqm, num_reads=4)), 4)
+
```
Comment (randomir) [inline tests/test_sampler.py:208]: It would be good to add some timeout tests.

**Review by randomir — COMMENTED** [review]
LGTM

File: tabu/src/tabu_search.cc (hunk 2 above)
Comment (hhtong) [reply tabu/src/tabu_search.cc:38]: All these checks are already done in sampler.py, so I figured that these were unneeded (Exceptions will never be thrown).

File: tabu/src/tabu_search.cc (hunk 3 above)
Comment (hhtong) [reply tabu/src/tabu_search.cc:120]: To summarize our conversation: Checking for positive/negative timeLimitInMilliSecs does not always tell us if we should/should not care about timeout, so this additional argument is needed for simpleTabuSearch().

File: tabu/src/tabu_search.cc (hunk 2 above)
Comment (randomir) [reply tabu/src/tabu_search.cc:38]: But what happens if someone uses C++ code directly (or via cython), and doesn't validate parameters?

Comment (hhtong) [reply tabu/src/tabu_search.cc:38]: Ah that's fair, I didn’t think we needed to take that into account. Will add these back!

File: tabu/sampler.py [hunk 5]
```diff
@@ -163,7 +175,8 @@ def sample(self, bqm, initial_states=None, initial_states_generator='random',
             # sanity check
             raise ValueError("unknown vartype")
 
-        return dimod.SampleSet.from_samples_bqm((samples, varorder), bqm=bqm)
+        return dimod.SampleSet.from_samples_bqm((samples, varorder), bqm=bqm, num_restarts=restarts)
```
Comment (randomir) [inline tabu/sampler.py:178]: Should be keep per-sample info in a sampleset field, or `info`? I can think of pros and cons for both approaches. @hhtong, @arcondello, thoughts?

File: tests/test_search.py [hunk 6]
```diff
@@ -99,7 +101,7 @@ def test_correctness(self):
     def test_concurrency(self):
 
         def search(timeout):
-            return tabu.TabuSearch([[1.0]], [1], 0, timeout).bestEnergy()
+            return tabu.TabuSearch([[1.0]], [1], 0, timeout, 1000000).bestEnergy()
```
Comment (randomir) [inline tests/test_search.py:104]: Clearer:
```
def search(timeout, restarts=int(1e6)):
    return tabu.TabuSearch([[1.0]], [1], 0, timeout, restarts).bestEnergy()
```

**Review by randomir — APPROVED** [review]
Looks good. I'd just like to discuss `num_restarts` field location before merging.

File: tabu/sampler.py (hunk 5 above)
Comment (arcondello) [reply tabu/sampler.py:178]: Unless it's a really unusual datatype or something, per-sample info should go in the `SampleSet.record`, that's the whole reason we have it! Especially for stuff like `num_restarts` which is just an int, I have a strong preference for using the record. Will let indexing, truncation, etc to work out of the box.

**Pull request merged**